route-rs-runtime/src/element/dec_ip_hop.rs

```diff
@@ -0,0 +1,96 @@
+use crate::element::Element;
+use smoltcp::wire::*;
+
+/*
+  DecIpv4Hop
+  Decrements the TTL of an IPv4 packet
+*/
+#[derive(Default)]
+pub struct DecIpv4HopLimit {}
+
+impl DecIpv4HopLimit {
+    pub fn new() -> DecIpv4HopLimit {
+        DecIpv4HopLimit {}
+    }
+}
+
+impl Element for DecIpv4HopLimit {
+    type Input = Ipv4Packet<Vec<u8>>;
+    type Output = Ipv4Packet<Vec<u8>>;
+
+    fn process(&mut self, mut packet: Self::Input) -> Self::Output {
+        packet.set_hop_limit(packet.hop_limit() - 1);
```

Comment:
What happens if the hop limit is currently the minimum allowed? Can we generate an invalid packet by doing this?

Reply:
One quick fix is to just never set this value below zero, and then leave it up to a classify element later to check ttl is greater than zero? The click codebase will be a decent guide

Reply:
I'm limiting this to 0 and passing the packet along unchanged. When we have more support for ICMP errors in the future we can generate an ICMP error instead.